### ISSUE TITLE
Self-hosted runner is running out of memory

### ISSUE THREAD

**Issue by nemanja899** [microsoft/AL-Go#781]
As per title , when configuring self-hosted runner in Al-Go settings, my virtual machine in Azure is running out of memory.

Anyway this is snipet of my Al-Go settings:
![image](https://github.com/microsoft/AL-Go/assets/34772552/dde42f0e-b490-4eac-a7ab-948fe74eb7ae)
It almost always downloads another image  for container even tho, I specified where to look in Al-Go settings.
My VM ends up with 10+ images with same repository name and with `<none>` as a name.
Please Help. Thank you

**Comment by jonaswre** [comment]
Hi, what VM size have you configured?

And do you mean memory or storage? 

If you meant storage  then my guess is 60 Days is too long. I think 3 Days is default. Caching every minor or new revision for 60 days is quite storage intensive.

**Comment by nemanja899** [comment]
Then what's the point running as self-hosed if image is downloaded multiple times a day. 
On Azure DevOps pipelines didn't had that problem, in one year images  were downloaded just a couple of times.

**Comment by freddydk** [comment]
Having 10 images built in one day seems wrong (unless you are targeting many countries)
As always, including the build log will help us see what's happening and why this is.

AL-Go caches images using BcContainerHelper, which probably is exactly what you did in Azure DevOps,
NextMinor and NextMajor are typically NOT cached, since these workflows are not run on a daily basis and the artifacts for these are almost certain to change.
Which artifacts are you building against? (what is the value of the artifact setting)
By default that is latest, but can be set to weekly or a specific version if you like (https://aka.ms/algosettings#artifact)

**Comment by jonaswre** [comment]
I didn't say they were downloaded multiple times a day each day.
I meant sometimes they are released very frequently. Especially after major releases. 

This is just something I realized from my own experience. Maybe the few times I noticed it, were just outlier's.

If I were you I would start at default value. And increase until you find a good balance between redownloading and storage. 

And one more... I think there was a bug regarding cleanup in an old version make sure your AL-Go is up-to-date.

**Comment by nemanja899** [comment]
This happens all the time
![Screenshot 2023-10-18 133157](https://github.com/microsoft/AL-Go/assets/34772552/d2490a68-fba3-41d4-948b-352b0999d627)
Had to delete them all in VM
And also every time there is a different tag new image is downloaded. Notice only one image version without none compare it to the new one in image below.
![image](https://github.com/microsoft/AL-Go/assets/34772552/36a03a3d-522d-4d68-bda9-c9fa6e897bc5)
Almost every time there is new tag, new image is pulled.

**Comment by jonaswre** [comment]
@freddy correct me if I'm wrong but the "none" images are only layers. They don't take up additional space right? 

Can you send a screenshot of how many space is actually being used? 

How many runners are on your VM and do you have any particular large repos? 

In the earlier days I encountered lots of people storing .app files in their repos.

Have you ever done that?

**Comment by nemanja899** [comment]
I deleted Images now only have one that u see in latest picture. "none" takes storage it fill up memory storage in a day. 
And yesterday had multiple images with "currentversion" repository name, had to go to VM to delete them also.
It's not that large repo and .app file is not stored in repo, but as an artifact when doing al-go Build action.

**Comment by jonaswre** [comment]
And you use which artifact in the Pipelines? 

Can you provide the following values after your cleanup and then after one and two runs?

Total space?
Total free space?
Space taken by the artifacts?  
Space taken by the docker images? Space taken by your work directories?

And after that maybe try to remove cacheImageName and cacheKeepDays and record the same values.

**Comment by nemanja899** [comment]
Total space : 127 GB
Total free space after removing images : 65.5 GB
There are two left, as shown in the latest image.
Now after running CI/CD
![image](https://github.com/microsoft/AL-Go/assets/34772552/522dacd9-6c4d-4f44-ab5f-f01b21cc0f30)

Again new image is pulled 

![image](https://github.com/microsoft/AL-Go/assets/34772552/f7fdcfd2-3222-4b76-9753-6886041821ae)

As you can see around 13GB is taken.

Artifacts size.
![image](https://github.com/microsoft/AL-Go/assets/34772552/533fd81a-643c-4886-a04e-836f47f472e1)


I think problem is with tag versions.

**Comment by freddydk** [comment]
I would add some more disk space and set keepdays to 3

Or use the artifact setting to use the same bc artifacts for longer time (daily, weekly or until you change it)

**Comment by jonaswre** [comment]
But if you rerun twice right after another size on disk shouldn't change.

I guess Microsoft is fixing a lot currently. 

It's by design. A new tagged version should be downloaded. If you don't want that, you can set the artifact to a specific version. 

But if you rerun twice right after another size on disk shouldn't change.

@freddydk if Microsoft publishes so many new images maybe AL-Go could get a max disk usage percentage setting. Or some sort of logic to only keep one minor on disk. This would at least not overflow smaller runners. 

But that's also something you could script on such a runner.

**Comment by freddydk** [comment]
Yeah, maybe we could do something in this area, to better work with smaller runners - but in the end, adding more disk space and setting KeepDays to 3 will cache all images that have been used for the last 3 days.

**Comment by jonaswre** [comment]
I would say in the long term it might be good to add some sort of "max disk usage" or "only keep one x versions of one minor". 
This would isolate runners from the frequency Microsoft publishes new versions. In theory Microsoft could currently overwhelm any runner if they just increase the frequency enough. 
We currently have around 200 GB storage for our runner that means 20 new releases in 3 Days.
Also unlikely still possible. And it should be fairly simple to append something to the cleanup.

**Comment by freddydk** [comment]
Better would be then to change the default artifact to daily instead of latest - this ensures that you will only grab one set of artifacts daily.

**Comment by jonaswre** [comment]
> Easier would be then to change the default artifact to daily instead of latest - this ensures that you will only grab one set of artifacts daily.

Is this already possible? 

This would still only solve it if you accept getting one artifact a day. 
People might still want to get the latest but only cache one version. 

But maybe people who need the latest but run small runners need to write there own code to clean it up. 
Because you would want that cleanup in general on a small runner no matter your configuration in the repos. Would be cool if the cleanup job looks for a specific file on the server an executes that. This would allow you to define cleanup scripts per runner/vm not per repo.

**Comment by freddydk** [comment]
By setting the artifact to "////daily" or "////weekly" you will minimize the number of new artifacts downloaded.
I would say that in most cases, a daily build is fine - and probably also the weekly.

**Comment by nemanja899** [comment]
@freddydk but when I use weekly build it throws an error in build
![image](https://github.com/microsoft/AL-Go/assets/34772552/411ddf4b-e70a-46af-83d5-c52f2626f242)
Al-Go settings
![image](https://github.com/microsoft/AL-Go/assets/34772552/921feb5c-6e51-45cb-b510-a1ef928b0ebb)
Not sure how to set -up for weekly artifact download
even tried
"artifact": "mcr.microsoft.com/sandbox/23.0.0.0/w1/<select parameter>

tried weekly in select parameter and then it thows an error 
![image](https://github.com/microsoft/AL-Go/assets/34772552/b646cff5-82c7-4469-a616-24660ca425bb)


parameter 'select'. should belong to the set "Latest,First,All,Closest,SecondToLastMajor,Current,NextMinor,NextMajor,Daily,Weekly"
I am really cofused  what to put in settings to ensure one artifact download weekly.
